### PR TITLE
Add aten::index_fill and its variants

### PR DESCRIPTION
**Pull request by yucai-intel**

### CODE REVIEW

File: yaml/xpu_functions.yaml [hunk 1]
```diff
@@ -86,6 +86,10 @@ supported:
   - index_add.out
   - index_add_
   - index_add
+  - index_fill_.int_Tensor
+  - index_fill_.int_Scalar
+  - index_fill.int_Tensor
+  - index_fill.int_Scalar
```
Comment (fengyuan14) [inline yaml/xpu_functions.yaml:92]:
```suggestion
  - index_fill_.int_Tensor
  - index_fill.int_Tensor
```

Comment (fengyuan14) [reply yaml/xpu_functions.yaml:92]: Scalar version operators should not be registered.

File: src/ATen/native/xpu/sycl/Indexing.cpp [hunk 2]
```diff
@@ -408,6 +408,99 @@ void index_add_kernel(
...
+  void operator()(
+      bool* dst,
+      bool* src,
+      int64_t dst_off,
+      int64_t src_off,
+      int64_t idx,
+      bool alpha) const {
+    dst[dst_off] = alpha;
+  }
+};
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/Indexing.cpp:434]: What's the difference of bool's specialization.

File: src/ATen/native/xpu/TensorAdvancedIndexing.cpp [hunk 3]
```diff
@@ -315,6 +315,71 @@ Tensor XPUNativeFunctions::index_add(
...
+        false,
+        "index_fill_(): Converting complex Scalar to non-complex type is not supported");
+  }
+
+  // Handle the case when `self` is 0-dim
+  Tensor self_nonzero_dim = (self.dim() == 0) ? self.unsqueeze(-1) : self;
+
+  dim = maybe_wrap_dim(dim, self.dim());
+  TORCH_CHECK(index.dim() <= 1, "Index has to be a vector/scalar");
+  index_func_meta_impl(self, self, dim, index, self, "index_fill");
```
Comment (fengyuan14) [inline src/ATen/native/xpu/TensorAdvancedIndexing.cpp:348]: No the calling in stock index_fill_ implementation. Please remove.

Comment (yucai-intel) [reply src/ATen/native/xpu/TensorAdvancedIndexing.cpp]: All the above issues have been corrected.

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**